### ISSUE TITLE
[Bug] Full node stopped working due to truncation panic

### ISSUE THREAD

**Issue by ppoliani** [aptos-labs/aptos-core#7982]
# 🐛 Bug

I've been running a full node (mainnet) for a couple of weeks now. However, it suddenly started failing due to a panic. More specifically, it panics when trying to truncate the state. More specifically it fails on this line

```rust
assert_eq!(position.to_postorder_index() + 1, num_frozen_nodes);
````

in the `truncate_transaction_accumulator` function.

**Stack trace/error message**
```
panicked at 'assertion failed: `(left == right)`
2023-04-29T08:41:30.169220Z [main] INFO storage/schemadb/src/lib.rs:168 Opened RocksDB. {"rocksdb_name":"state_merkle_db"}
  left: `263249952`,
 right: `263249957`', storage/aptosdb/src/utils/truncation_helper.rs:156:5'''
2023-04-29T08:41:30.169250Z [main] INFO storage/aptosdb/src/lib.rs:426 State K/V DB is not enabled!
backtrace = '''
2023-04-29T08:41:30.169256Z [main] INFO storage/aptosdb/src/lib.rs:428 Opened AptosDB (LedgerDB + StateMerkleDB). {"ledger_db_path":"/xxxxx/xxxxxxx/.aptos/db/ledger_db","state_merkle_db_path":"/xxxx/xxxx/.aptos/db/state_merkle_db","time_ms":"292"}
   0:     0x5562163ffb46 - backtrace::backtrace::libunwind::trace::hfdb078f3c7dfc4ae
                               at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/backtrace/libunwind.rs:93:5
                           backtrace::backtrace::trace_unsynchronized::h111e07def32e76e7
2023-04-29T08:41:30.169365Z [main] INFO storage/aptosdb/src/pruner/db_pruner.rs:18 state_merkle_pruner initialized. {"min_readable_version":130623122}
                               at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/backtrace/mod.rs:66:5
2023-04-29T08:41:30.169460Z [main] INFO storage/aptosdb/src/pruner/db_pruner.rs:18 state_merkle_pruner initialized. {"min_readable_version":51624912}
                           backtrace::backtrace::trace::hd3d5670e1160f40d
                               at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/backtrace/mod.rs:53:14
2023-04-29T08:41:30.169588Z [main] INFO storage/aptosdb/src/state_store/mod.rs:289 Start syncing databases... {"overall_commit_progress":131624982}
   1:     0x5562163e9481 - backtrace::capture::Backtrace::create::h126140a7b8886788
                               at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/capture.rs:176:9
                           backtrace::capture::Backtrace::new::h869bfdd5d4d1cdb2
                               at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/capture.rs:140:22
   2:     0x5562144ba544 - aptos_crash_handler::handle_panic::h9b058a1a75ad7654
2023-04-29T08:41:30.169607Z [main] INFO storage/aptosdb/src/state_store/mod.rs:308 Start truncation... {"ledger_commit_progress":131624984}
                               at /code/aptos-core/crates/crash-handler/src/lib.rs:37:38
                           aptos_crash_handler::setup_panic_handler::{{closure}}::h35ce253fcae91f1b
                               at /code/aptos-core/crates/crash-handler/src/lib.rs:29:9
   3:     0x5562165a5e49 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hbb63890df2a28e48
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/alloc/src/boxed.rs:2001:9
                           std::panicking::rust_panic_with_hook::h4b1447a24e3e94f8
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:692:13
   4:     0x5562165a5bc7 - std::panicking::begin_panic_handler::{{closure}}::h8701da9995a3820c
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:579:13
   5:     0x5562165a44ac - std::sys_common::backtrace::__rust_end_short_backtrace::hb696c5ed02a01598
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/sys_common/backtrace.rs:137:18
   6:     0x5562165a58e2 - rust_begin_unwind
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:575:5
   7:     0x5562165c86a3 - core::panicking::panic_fmt::h8aa706a976963c88
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/panicking.rs:65:14
   8:     0x5562165c8a3b - core::panicking::assert_failed_inner::h3c3fae67130f6d6f
   9:     0x55621476ed07 - core::panicking::assert_failed::hfa6f8b380a5352fd
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/panicking.rs:203:5
  10:     0x5562147def25 - aptos_db::utils::truncation_helper::truncate_transaction_accumulator::h09a5a83a93b9df6e
                               at /code/aptos-core/storage/aptosdb/src/utils/truncation_helper.rs:156:5
                           aptos_db::utils::truncation_helper::truncate_ledger_db_single_batch::hcf7560c0e1c32cbc
                               at /code/aptos-core/storage/aptosdb/src/utils/truncation_helper.rs:193:5
                           aptos_db::utils::truncation_helper::truncate_ledger_db::h5492e333e0f7e99b
                               at /code/aptos-core/storage/aptosdb/src/utils/truncation_helper.rs:51:9
  11:     0x5562147af638 - aptos_db::state_store::StateStore::new::h56fdf5c721e2cc4f
                               at /code/aptos-core/storage/aptosdb/src/state_store/mod.rs:314:17
  12:     0x5562147b685d - aptos_db::AptosDB::new_with_dbs::hfa19977ece0d4f0b
                               at /code/aptos-core/storage/aptosdb/src/lib.rs:291:36
  13:     0x556213fbfc02 - aptos_db::AptosDB::open::h9a68de3c4782cf84
                               at /code/aptos-core/storage/aptosdb/src/lib.rs:343:26
                           aptos_node::storage::initialize_database_and_checkpoints::h7fd458592e444d9d
                               at /code/aptos-core/aptos-node/src/storage.rs:91:20
  14:     0x556213fc4135 - aptos_node::setup_environment_and_start_node::h84365ff19f6c24df
                               at /code/aptos-core/aptos-node/src/lib.rs:362:9
  15:     0x556213fc3944 - aptos_node::start::h7f5431c18d13f978
                               at /code/aptos-core/aptos-node/src/lib.rs:187:9
  16:     0x556213fc21fc - aptos_node::AptosNodeArgs::run::h52287ba0862575f7
                               at /code/aptos-core/aptos-node/src/lib.rs:128:13
  17:     0x556213eee3b0 - aptos_node::main::h4cf5b732ed61bad5
                               at /code/aptos-core/aptos-node/src/main.rs:19:5
  18:     0x556213eebfb6 - core::ops::function::FnOnce::call_once::ha37e11ed55fd6150
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/ops/function.rs:251:5
                           std::sys_common::backtrace::__rust_begin_short_backtrace::h99fdddf089890790
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/sys_common/backtrace.rs:121:18
  19:     0x556213eee4ec - std::rt::lang_start::{{closure}}::h34b3b855645faeb8
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/rt.rs:166:18
  20:     0x55621659632b - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h8cbb48ae40ddb046
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/ops/function.rs:286:13
                           std::panicking::try::do_call::h92db802eb38b49b7
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:483:40
                           std::panicking::try::ha8949d2082cf3644
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:447:19
                           std::panic::catch_unwind::h5e34c1f8a5992ed9
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panic.rs:137:14
                           std::rt::lang_start_internal::{{closure}}::hea52a0bb3f8ff16a
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/rt.rs:148:48
                           std::panicking::try::do_call::h5bc358faf3d68a8b
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:483:40
                           std::panicking::try::h675304212928379d
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:447:19
                           std::panic::catch_unwind::h7ce3ad349ed5c844
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panic.rs:137:14
                           std::rt::lang_start_internal::hcd7e45acd25ab5ab
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/rt.rs:148:20
```

## System information

**Please complete the following information:**
aptos-node-v1.3.2
rustc 1.68.2
Ubuntu 22.10 x64

**Comment by grao1991** [comment]
I think you hit a known bug that has already been fixed in the v1.4.

**Comment by grao1991** [comment]
I think now your db is at an inconsistent state. It will require either resync your node, or if you can build from source, we can probably try to do some manual operation to fix it.

**Comment by ppoliani** [comment]
Hey @grao1991, as a matter of fact I build the binaries from the source code. I did circumvent this issue by commenting out the call to `truncate_ledger_db` [here](https://github.com/Apocentre/aptos-core/blob/main/storage/aptosdb/src/state_store/mod.rs#L363). It worked pretty well for several hours until the node started failing as described here https://github.com/aptos-labs/aptos-core/issues/7985.

I'm not sure if these two events are related in any possible way.

FYI, I built the binaries on top of the `aptos-release-v1.3` branch.

**Comment by grao1991** [comment]
Tl;Dr, I think the other issue you posted is probably the root cause.

It's not easy to tell when your db enters a bad state. I tend to believe your state data got corrupted first and you got a panic (like the other issue you posted). Then your node restarted and noticed the ledger commit progress is ahead and needs to be truncated, then hit this bug.

I'm not 100% sure there are no corner cases, but skipping the ledger truncation should be an acceptable way to bypass this issue, because as long as the chain is not forked, your node should write the exact same data for any given version..

**Comment by ppoliani** [comment]
@grao1991 I'll close this since you mentioned that this edge case is fixed in version 1.4. Thanks again for your input, much appreciated :)
